### PR TITLE
database objects have (default) template context

### PR DESCRIPTION
**Pull request by danfrankj**
@matthewwardrop 

This change would let us set a property `template_context` on database objects that would contain "default context". This default context would then be overwritten by anything explicitly passed 

```
In [1]: ap.presto.template_context = {'current_ds': lambda : datetime.date.today().isoformat()}

In [2]: %%presto
   ...: SELECT '{{ current_ds() }}'
   ...:
presto: Query: Complete after 0.65 sec on 2017-09-27. CAVEATS: Loaded from cache.
Out[2]:
        _col0
0  2017-09-27
``` 

This will be useful so that we can set up a default template context when these ducts are initialized and then you don't have to populate your local namespace with random variables / functions.

### CODE REVIEW

**Comment by danfrankj** [comment]
@matthewwardrop thoughts?

**Comment by matthewwardrop** [comment]
@danfrankj Thanks for the contribution, Dan!

I'm of mixed minds about including this sort of feature. At this stage, apart from connection configuration, there is no "magic" state which can affect how code works in differing contexts. In fact, so far, the entire point of omniduct has been to abstract away details of the implementation so that every database duct operation works atomically. This feature seems to me to be something that could mess with that... where the order in which code cells run can change Duct state, and hence cause omniduct to run differently depending on code execution order.

I can see why this feature can be useful, I'm just not sure it's worth the sacrifice of atomicity. What do you think?

**Comment by danfrankj** [comment]
@matthewwardrop as per our discussion in person, I think the remaining asks for this PR were 

- [ ] warning if default context and context passed in conflict
- [ ] public method to set_default_context 
- [ ] public method to update_default_context 

Anything else you had in mind?

**Comment by matthewwardrop** [comment]
Hi @danfrankj ,

Thanks for convincing me about this. I think this feature will be useful :).

I think the best way to add it is to add a `_template_context` attribute to the constructor (under line 76), and to allow values for them to specified by popping values passed as `templates` and `template_context` to the constructor in `kwargs` (defaulting both to the empty dictionary). With that, I think I'm fine with the direct modification of the `template_context` attribute as you're doing presently, and so we wouldn't need methods for updating/setting the template context.

On top of that, as you note, I think having the warning will be helpful to prevent confusing things happening when people copy and paste code.

**Comment by danfrankj** [comment]
@matthewwardrop updated - PTAL

File: omniduct/databases/base.py [hunk 1]
```diff
@@ -71,7 +79,8 @@ def __init__(self, *args, **kwargs):
         '''
         Duct.__init_with_kwargs__(self, kwargs, port=self.DEFAULT_PORT)
 
-        self._templates = {}
+        self._templates = kwargs.pop('templates')
```
Comment (matthewwardrop) [inline omniduct/databases/base.py]: Both of these should have a default value supplied of '{}'.

File: omniduct/databases/base.py [hunk 2]
```diff
@@ -35,10 +35,18 @@ def render_statement(method, self, statement, *args, **kwargs):
...
-    if template:
+    if kwargs.pop('template', False):
+        template_context = {}
+        template_context.update(self._template_context)
+        kwarg_context = kwargs.pop('template_context', {})
+        template_context.update(kwarg_context)
+        intersection = set(self._template_context.keys()) & set(kwarg_context.keys())
+        if intersection:
+            logger.warning(
+                'Duct default context conflicts with the context passed in '
```
Comment (matthewwardrop) [inline omniduct/databases/base.py]: I think this logging message has the potential to confuse some people, as templates are never mentioned.

How about: "The following default template context keys have been overridden by the local context: {}."

**Comment by matthewwardrop** [comment]
@danfrankj Made some small corrections/suggestions, once they've been resolved, merge at will.

**Pull request merged**